onnxruntime/core/providers/cuda/tensor/gather_elements.cc

```diff
@@ -1,101 +1,216 @@
+                        int64_t& new_axis, int64_t& new_rank, int64_t& input_stride_along_axis,
+                        TArray<int64_t>& masked_input_strides, TArray<fast_divmod>& indices_fdms) {
+  size_t rank = input_shape.size();
+  // Reverse for better calculation.
+  std::reverse(input_shape.begin(), input_shape.end());
+  std::reverse(indices_shape.begin(), indices_shape.end());
+  size_t reverse_axis = rank - 1 - static_cast<size_t>(axis);
+  size_t curr = 0, next = 0;
+  while (curr < reverse_axis && CanSkip(input_shape, indices_shape, curr)) ++curr;
+  if (curr < reverse_axis) {
+    if (curr > 0) Move(input_shape, indices_shape, curr, 0);
+    next = curr + 1;
+    curr = 0;
+    while (next < reverse_axis) {
+      if (!CanSkip(input_shape, indices_shape, next)) {
+        if (CanMerge(input_shape, indices_shape, next, curr)) {
+          Merge(input_shape, indices_shape, next, curr);
+        } else {
+          ++curr;
+          if (curr != next) Move(input_shape, indices_shape, next, curr);
+        }
+      }
+      ++next;
+    }
+  }
+
+  if (curr == reverse_axis) {
+    if (curr > 0) Move(input_shape, indices_shape, curr, 0);
+    curr = 0;
+  } else {
+    ++curr;
+    // next is now the reverse_axis
+    if (curr != next) Move(input_shape, indices_shape, next, curr);
+  }
+  size_t new_reverse_axis = curr;
+  next = reverse_axis + 1;
+  while (next < rank && CanSkip(input_shape, indices_shape, next)) ++next;
+  if (next < rank) {
+    ++curr;
+    if (curr != next) Move(input_shape, indices_shape, next, curr);
+    ++next;
+    while (next < rank) {
+      if (!CanSkip(input_shape, indices_shape, next)) {
+        if (CanMerge(input_shape, indices_shape, next, curr)) {
+          Merge(input_shape, indices_shape, next, curr);
+        } else {
+          ++curr;
+          if (curr != next) Move(input_shape, indices_shape, next, curr);
+        }
+      }
+      ++next;
+    }
+  }
+
+  new_rank = curr + 1;
+  new_axis = static_cast<int64_t>(new_rank - 1 - new_reverse_axis);
+  input_shape.resize(new_rank);
+  std::reverse(input_shape.begin(), input_shape.end());
+  indices_shape.resize(new_rank);
+  std::reverse(indices_shape.begin(), indices_shape.end());
+
+  // Set stride along axis to 0 so we don't need IF statement to check in kernel.
+  TensorPitches masked_input_strides_vec(input_shape);
+  input_stride_along_axis = masked_input_strides_vec[new_axis];
+  masked_input_strides_vec[new_axis] = 0;
+  int32_t new_rank_32bit = static_cast<int32_t>(new_rank);
+  masked_input_strides.SetSize(new_rank_32bit);
+  indices_fdms.SetSize(new_rank_32bit);
+  TensorPitches indices_strides(indices_shape);
+  for (auto i = 0; i < new_rank_32bit; ++i) {
+    masked_input_strides[i] = masked_input_strides_vec[i];
+    indices_fdms[i] = fast_divmod(gsl::narrow_cast<int>(indices_strides[i]));
+  }
+}
+
+// GatherElementsGrad needs atomic_add which supports float types only, so use half, float and double for 16, 32, and 64
```

Comment:
while looks like this func not only support float, but also int8.

Reply:
All these Ops actually doesn't care about the real data type, but care about the size of data. This comment is actually explaining why case sizeof(int8_t) returns TensorProto_DataType_INT8, but for int16, int32 and int64, I didn't use corresponding TensorProto_DataType_INT type, but use TensorProto_DataType_FLOAT types. The reason is if I use TensorProto_DataType_INT types, it will not work for GatherElementsGrad as atomic_add will be called. But we don't have this problem for int8 because only GatherElements/ScatterElements support this type, GatherElementsGrad doesn't support it.

Reply:
This is not a new idea. Most of our data-transfer kernels are doing this way, such as Transpose, Tile, etc. Transpose uses int types, while Tile use float types, they are doing fine. Since there is no real data compute for these kernels, it's just data assignment from memory to memory, I don't think any bit in the data will be changed.